Introduce Transaction Subsystem Fuzzing and Common Fuzzer Interface

robertDurst opened the pull request:
# Some Background

Fuzzing is a test strategy that performs automated binary executions with randomly generated/mutated inputs, monitoring for program crashes. It is useful for increasing test coverage by discovery of obscure, unthought-of, and interesting edge cases.

The basic process works as so:
```
**Input**: an initial corpus of well formed data
1. Inputs fed into a program
2. Fuzzer derives more inputs (based on some gathered metrics like unique execution path)
3. Repeat
**Output**: inputs that caused crashes
```
There are many fuzz testing tools out there such as [libFuzzer](https://llvm.org/docs/LibFuzzer.html), [AFL](http://lcamtuf.coredump.cx/afl/), and [Peach Fuzzer](https://www.peach.tech/products/peach-fuzzer/). For the time being, I have implemented a test harness to work with AFL.

# Description

This builds on #2155, introducing a transaction subsystem fuzzer.

A significant portion of the work here was integrating the introduced transaction fuzzer with [AFL's persistent mode](https://lcamtuf.blogspot.com/2015/06/new-in-afl-persistent-mode.html). Persistent mode allows for multiple executions of the binary in a single long-lived process, significantly improving executions per second since the expensive state setup can be limited to every 1,000th, or even 1,000,000th execution. Areas where determinism improved include:
* foregoing signature verification (we aren't breaking ED25519 here)
* clearing caches
* seeding randomness

This PR also defines a common interface for fuzzing various subsystems, allowing one to utilize a CLI flag to switch between `Overlay` and `Transaction` fuzzing.

Finally, this PR includes an option to specify `process_id`, allowing for easy parallelization of fuzzing (as supported by AFL). 

Further progress of #1376 

# Future Work

### Generating transactions with more operations
To demonstrate functionality, I restricted the fuzzer to a subset of operations: `ACCOUNT_MERGE` and `PAYMENT` of native assets. Running against protocol 4, the fuzzer discovered the [double merge bug described here](https://www.stellar.org/developers/stellar-core/software/security-protocol-release-notes.html#v061b-not-widely-released-2017-04-06). This set will be expanded as I extend the fuzz test harness methodically, focusing first on discovery of the rest of the [historical bugs](https://www.stellar.org/developers/stellar-core/software/security-protocol-release-notes.html#v061b-not-widely-released-2017-04-06) in chronological order.

### Improving Fuzzer Efficiency
I also spent some time thinking about how to meaningfully and unobstructively limit the *fuzz space*. For example, I use a small set of pre-generated accounts for operations. Since I do this, and cause a disconnect between what the fuzzer fuzzes and what is actually fed to Stellar-core, cpu time spent incrementing and flipping bits in the `AccountID` property of XDR's is wasted time. Thus, I feed the fuzzer an initial corpus of operations with all bytes of the `AccountID`  zero'd out except for the 32nd byte to nudge it in the right direction.

***

### Note for Reviewers

Dependent on and rebased against #2155.

# Checklist
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests

MonsieurNicolas commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1288,6 +1289,21 @@ LedgerTxnRoot::Impl::~Impl()
 
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+void
+LedgerTxnRoot::Impl::clearBestOfferAndEntryCaches()
```
just call this method something more generic, like `resetForFuzzer`

MonsieurNicolas commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -1903,6 +1919,11 @@ LedgerTxnRoot::Impl::getNewestVersion(LedgerKey const& key) const
     putInEntryCache(key, entry, LoadType::IMMEDIATE);
+
+#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION
+    mDatabase.clearPreparedStatementCache();
```
why isn't this done in `resetForFuzzer`?

MonsieurNicolas commented on src/main/CommandLine.cpp:
```diff
@@ -159,6 +159,20 @@ fileNameParser(std::string& string)
+processIDParser(uint& num)
+{
+    return clara::Opt{num, "PROCESS-ID"}["--process_id"](
+        "for spawning multiple instances in fuzzing paralellization");
```
sp: "parallelization"

MonsieurNicolas commented on src/main/CommandLine.cpp:
```diff
@@ -159,6 +159,20 @@ fileNameParser(std::string& string)
+clara::Opt
+isOverlayFuzzer(bool& overlayFuzz)
+{
+    return clara::Opt{overlayFuzz}["--overlay_fuzz"](
```
wouldn't it be cleaner to have something like
`fuzz --mode=overlay` or `fuzz --mode=tx` ; that way we can add as many fuzzers as we want

MonsieurNicolas commented on src/test/fuzz.cpp:
```diff
@@ -1,194 +0,0 @@
-// under the Apache License, Version 2.0. See the COPYING file at the root
-// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
-
-#include "util/asio.h"
```
no need to move files to a different folder, it just adds noise to the tree

MonsieurNicolas commented on src/test/fuzz/FuzzTransaction.cpp:
```diff
@@ -0,0 +1,66 @@
+
+    TransactionResultCode code;
+    {
+        LedgerTxn ltxFeeProc(ltx);
```
I don't think we need this (processing fees)

MonsieurNicolas commented on src/test/fuzz/FuzzTransaction.cpp:
```diff
@@ -0,0 +1,66 @@
+        // after application, clear cache so that they don't persist between
+        // executions and confuse the fuzzer, causing it to think some Tx with
+        // [op0...opN] can result in multiple different execution paths.
+        app.getLedgerTxnRoot().clearBestOfferAndEntryCaches();
```
I thought the `reset` code would be done *before* running a test (at the same time than reseeding)

MonsieurNicolas commented on src/test/fuzz/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,192 @@
+using namespace stellar;
+
+// for a given XDRInputFileStream, continuously try to read and inject an XDR
+template <typename T, typename Injection>
```
you don't need templating for this

MonsieurNicolas commented on src/test/fuzz/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,192 @@
+    while (tryRead(in, inp))
+    {
+        Fuzzer::seedRandomness();
+        LOG(INFO) << "Fuzzer injecting message ";
```
I don't think logging is a good idea in the main fuzz loop

MonsieurNicolas commented on src/test/fuzz/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,192 @@
+
+            try
+            {
+                fuzztest::attemptTx(txFramePtr, *app);
```
this should just be a regular method, not calling statics like that

MonsieurNicolas commented on src/test/fuzz/deserialize.cpp:
```diff
@@ -0,0 +1,94 @@
+{
+
+TransactionFramePtr
+deserializeFuzzTransactionInput(FuzzTransactionInput const& inp,
```
I don't think you need this machinery:
if you generate test accounts that have a "public key" 000000XX, then the mapping happens implicitly and you don't have to scan the XDR to search/replace

robertDurst replied:
Ok, I see... at first I considered the impossibility of predictability generating ED25519 keypairs, but then remembered that we aren't checking signatures here, so secret keys don't matter!

MonsieurNicolas commented on src/test/fuzz/fuzz.cpp:
```diff
@@ -0,0 +1,190 @@
+namespace xdr
+{
+template <>
+typename std::enable_if<xdr_traits<stellar::PublicKey>::is_union>::type
```
do you need this?

MonsieurNicolas commented on src/test/fuzz/fuzz.cpp:
```diff
@@ -0,0 +1,190 @@
+    typename xdr_traits<stellar::PublicKey>::discriminant_type v;
+
+    // Just pick a random case
+    uint32_t n;
```
not sure I understand what you're trying to do here

MonsieurNicolas commented on src/xdr/Fuzz-test.x:
```diff
@@ -1,15 +1,15 @@
-void fuzz(std::string const& filename, el::Level logLevel,
-          std::vector<std::string> const& metrics);
-void genfuzz(std::string const& filename);
+struct FuzzTransactionInput
```
you don't need a custom type for this (and the xdr file): all you need to do is generate random operations

robertDurst commented:
More or less addressed everything above. Some of the hacky-est parts here will be benefited by the work being done in #2179. Specifically, I created a method called `resetAndApplyForFuzzer` that calls `resetResults` and `applyOperations` on a `TransactionFrame`, in order to bypass fee processing. I also manually construct a transaction envelope, replicating some of the code in `TestAccount` since I only have public keys and only care about the public keys of my pregenerated accounts.

Will move this to **Ready for Review** once the PR mentioned above and #2155 are complete.

vogel commented on src/ledger/LedgerTxn.cpp:
```diff
@@ -83,7 +83,8 @@ EntryIterator::getImpl() const
     return mImpl;
 }
 
-EntryIterator& EntryIterator::operator++()
```
looks like you are using different version of clang-format that everyone

robertDurst replied:
Fixing, it looks like I was using 7 instead of 5 -- thanks!

vogel commented on src/test/fuzz.cpp:
```diff
@@ -1,194 +1,186 @@
+        return FuzzerMode::OVERLAY;
+    }
+
+    if (iequals(fuzzerMode, "tx") || iequals(fuzzerMode, "transaction"))
```
that seems not necessary because of default return value

robertDurst replied:
Good point, done.

robertDurst commented:
Rebased against master. With #2179 merged, going to make some changes here based on that PR and undraft later today.

robertDurst commented:
This PR is now ready for review! 

**Areas of uncertainty**:
* currently for initial setup, I pregenerate accounts. As it is currently implemented, I generate 16 public keys (reasons explained in comments) and add these to the ledger via creation of `LedgerEntry` objects. As I played around a bit with order book initial state setup, I do not think this elegantly scales to more complex state setup. Originally I had a ``FuzzAccount` (subset of `TestAccount`) object and utilized `TxTests`  which made this stuff a bit easier, however this comes with some overhead such as many methods having a `REQUIRE` macro that I do not want/can't use. Do note, this is only an issue for initial setup -- now that I pregenerate accounts within a predicable range, I can directly use fuzzed XDR's as input.
* Not really commentary on the current PR, but related. I have discussed with Nicolas what makes a good initial corpus for fuzzing. Ideally this would be running `genFuzz` many, many (1000's/10000's) of times and then using AFL's corpus minimization tool `afl-cmin`. Thus, there is no reason to run this fuzz initial corpus generator every single time I want to `fuzz`, may be worth having a repo or home for large, minimized initial corpuses.

robertDurst commented:
Appears I broke something. Not fuzzing correctly anymore. Will work on fixing tomorrow.

robertDurst commented:
My apologies for not finding the above bug before opening for review. Tested and everything works again as expected. Should be good for review again.

MonsieurNicolas commented on src/test/fuzz.cpp:
```diff
@@ -227,4 +227,16 @@ genfuzz(std::string const& filename)
         }
     }
 }
+// default "TRANSACTION" if unrecognized or unspecified
```
no silent failures please: if not recognized -> help

MonsieurNicolas commented on src/test/fuzz.cpp:
```diff
@@ -1,144 +1,83 @@
-
-bool
-tryRead(XDRInputFileStream& in, StellarMessage& m)
+template <typename XDR_TYPE, typename Predicate>
```
it would be cleaner to have this functionality attached to the respective fuzz class, that way you can share the validation code between genfuzz and regular fuzzing.
And if you do it like this, you won't need templates

MonsieurNicolas commented on src/test/fuzz.cpp:
```diff
@@ -147,86 +86,42 @@ fuzz(std::string const& filename, el::Level logLevel,
 #endif // AFL_LLVM_MODE
     {
-        XDRInputFileStream in(MAX_MESSAGE_SIZE);
+        XDRInputFileStream in(MAX_XDR_SIZE);
```
`MAX_XDR_SIZE` doesn't mean anything: this is the maximum xdr object size

MonsieurNicolas commented on src/test/fuzz.cpp:
```diff
@@ -1,144 +1,83 @@
-#define INITIATOR 1
-#define ACCEPTOR 0
+#define PERSIST_MAX 1000000
+#define MAX_XDR_SIZE 0x1000000
```
I don't think you should use this value `0x1000000` for transactions (way too big)

robertDurst replied:
Ah yes... was just reading how someone 8x'd their speed by limiting to 50kb

robertDurst replied:
Averaged XDR size of 100k `path_payment` only XDR's in the range of 1-3 ops and got `294 bytes`. So I think `0x500` should be fine.

MonsieurNicolas commented on src/test/fuzz.cpp:
```diff
@@ -117,10 +116,32 @@ genfuzz(std::string const& filename, FuzzerMode fuzzerMode)
+                    op.body.paymentOp().asset.type() == ASSET_TYPE_NATIVE);
+        };
+
+        auto isDesiredXDR = [predicate](xdr::xvector<Operation>& ops) {
```
rework this: this inefficient and convoluted

MonsieurNicolas commented on src/test/fuzz.cpp:
```diff
@@ -117,10 +116,32 @@ genfuzz(std::string const& filename, FuzzerMode fuzzerMode)
     {
+        // TODO: allow for more operations, not just account merge and
+        // native payments
+        auto predicate = [](Operation& op) {
```
this predicate should be shared (see previous comment on genfuzz)

MonsieurNicolas commented on src/test/fuzz.cpp:
```diff
@@ -117,10 +116,32 @@ genfuzz(std::string const& filename, FuzzerMode fuzzerMode)
     }
     case FuzzerMode::TRANSACTION:
     {
+        // TODO: allow for more operations, not just account merge and
```
remove TODO from codebase

MonsieurNicolas commented on src/test/fuzz.cpp:
```diff
@@ -155,3 +155,20 @@ getFuzzerModefromString(std::string const& fuzzerMode)
+    // 31st indexed byte. Furthermore, utilize only the first 4 bits of the 32nd
+    // byte, allowing us to generate these accounts against a pregeneration of
+    // such accountID's.
+    t.ed25519().at(31) = autocheck::generator<uint8_t>()(10) % 16;
```
I thought the generator for scalars was taking the max value?

MonsieurNicolas replied:
also do not use hardcoded numbers for the number of accounts

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -37,6 +37,7 @@ getFuzzConfig(int instanceNumber)
     cfg.QUORUM_INTERSECTION_CHECKER = false;
     cfg.PREFERRED_PEERS_ONLY = false;
     cfg.RUN_STANDALONE = true;
+    cfg.BUCKET_DIR_PATH = "Bucket" + std::to_string(instanceNumber);
```
you should not need this

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -114,6 +115,104 @@ class FuzzTransactionFrame : public TransactionFrame
+    // now we create 16 accounts, or enough to fill the first 4 bits such that
+    // we have a pregenerated account for the last 4 bits of the 32nd byte of a
+    // public key
+    for (int i = 0; i < 16; ++i)
```
do not use hardcoded numbers

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -114,6 +115,104 @@ class FuzzTransactionFrame : public TransactionFrame
+        newAccount.accountID = publicKey;
+        newAccount.seqNum = 0;
+        // to create "interesting" balances we utilize powers of 2
+        newAccount.balance = pow(2, i);
```
do not use `pow` for integers (shift operator)

robertDurst commented:
Ok, should be good for another review.

jonjove commented on src/test/FuzzerImpl.cpp:
```diff
@@ -48,6 +48,72 @@ seedRandomness()
+        app.getLedgerTxnRoot().resetForFuzzer();
+        app.getDatabase().clearPreparedStatementCache();
+
+        // reset results of operations
```
This entire block looks nearly duplicate to `TransactionFrame::resetResults`. Can you refactor to share code?

jonjove commented on src/test/FuzzerImpl.cpp:
```diff
@@ -90,6 +90,7 @@ class FuzzTransactionFrame : public TransactionFrame
         SignatureChecker signatureChecker{
             ltx.loadHeader().current().ledgerVersion, getContentsHash(),
             mEnvelope.signatures};
+        loadSourceAccount(ltx, ltx.loadHeader());
```
Probably good to have a comment explaining why this is here.

jonjove commented on src/test/FuzzerImpl.cpp:
```diff
@@ -114,6 +115,104 @@ class FuzzTransactionFrame : public TransactionFrame
+            return;
+
+        // construct transaction
+        auto txFramePtr = ::FuzzTransactionFrame::create(mSourceAccountID, ops,
```
Nitpick: `::FuzzTransactionFrame` -> `FuzzTransactionFrame`

jonjove commented on src/test/FuzzerImpl.cpp:
```diff
@@ -114,6 +115,104 @@ class FuzzTransactionFrame : public TransactionFrame
+
+            // we need an extra db cache clear here since the db prepared
+            // statement cache may persist between tryRead loops
+            mApp->getDatabase().clearPreparedStatementCache();
```
Why is this needed? You are already clearing this in `attemptApplication`.

robertDurst replied:
I cannot quite figure out why it is needed -- for some reason, calling this before the `XDRStream` reads another `XDR` in `tryRead` makes the stability go from `91% --> 95%` even though I am fairly confident the two are wholly unrelated.

robertDurst replied:
*Another general observation:* turns out that the first `clearPreparedStatementCache` has generally no affect on stability, given the existence of this, the second `clearPreparedStatementCache` call. This makes sense given that it clears the cache after the `tx attempted apply`, before another loop iteration, and before `XDRStream`.

MonsieurNicolas replied:
`resetTxInternalState` should call `clearPreparedStatementCache`

MonsieurNicolas replied:
I suspect you need to call `clearPreparedStatementCache` after `ltx` goes out of scope just to be sure things are reset the way you want

jonjove commented on src/test/FuzzerImpl.cpp:
```diff
@@ -114,6 +115,104 @@ class FuzzTransactionFrame : public TransactionFrame
+        {
+            throw;
+        }
+        catch (TransactionResultCode const& err)
```
I don't think we ever actually throw a `TransactionResultCode`. How is this block reached? I think to catch internal errors correctly, you need to interpret any exception other than `InvariantDoesNotHold` as an internal error. Based on the comment in the next block, this might be a non-trivial change requiring some validity checks to filter out other error sources.

robertDurst replied:
Ok, so you are correct, TransactionResultCode is never thrown. It was thrown in the original version of the PR, but is no longer, so that has been removed. 

Also, I figured out that the **other** exceptions, such as the ones I note in the comment are due to me removing a `checkValid` call. I originally removed this because I though it just checked signatures and such -- however it checks if operations are well-formed as well, and as far as I can tell, the other exceptions being caught were all mal-formed XDR exceptions such as `terminate called after throwing an instance of 'xdr::xdr_wrong_union'`.

My solution for this is re-including the `checkValid` and if it fails for any operation, returning early, and not *"applying the transaction"*.

jonjove commented on src/test/FuzzerImpl.cpp:
```diff
@@ -216,6 +214,39 @@ TransactionFuzzer::inject(stellar::XDRInputFileStream& in)
+        auto numops = autocheck::generator<uint>()(6);
+        for (int i = 0; i < numops; ++i)
+        {
+            Operation op = gen(14);
```
Why `14` here and two lines later? Try to avoid magic numbers whenever possible

jonjove commented on src/test/FuzzerImpl.cpp:
```diff
@@ -314,4 +330,37 @@ OverlayFuzzer::inject(stellar::XDRInputFileStream& in)
+    autocheck::generator<StellarMessage> gen;
+    try
+    {
+        StellarMessage m(gen(10));
```
Same comment about magic numbers.

jonjove commented on src/test/FuzzerImpl.cpp:
```diff
@@ -216,6 +214,39 @@ TransactionFuzzer::inject(stellar::XDRInputFileStream& in)
+
+        out.writeOne(ops);
+    }
+    catch (xdr::xdr_bad_discriminant const&)
```
Is it possible to change the above so `xdr_bad_discriminant` never happens?

robertDurst replied:
I actually do not believe this is triggered since these xdr's are generated by `autocheck`, which knows enough to generate valid structures.

jonjove commented on src/test/FuzzerImpl.cpp:
```diff
@@ -314,4 +330,37 @@ OverlayFuzzer::inject(stellar::XDRInputFileStream& in)
+
+        out.writeOne(m);
+    }
+    catch (xdr::xdr_bad_discriminant const&)
```
Same question about `xdr_bad_discriminant`.

robertDurst commented on src/test/fuzz.h:
```diff
@@ -1,15 +1,23 @@
 #pragma once
 
-// Copyright 2015 Stellar Development Foundation and contributors. Licensed
+// Copyright 2019 Stellar Development Foundation and contributors. Licensed
```
So I asked Marta and the consensus seems to be, for any new files use this year, otherwise don't touch -- switching all the touched dates back.

marta-lokhova commented on src/test/FuzzerImpl.h:
```diff
@@ -0,0 +1,43 @@
+  public:
+    TransactionFuzzer(int numAccounts, int processID);
+    void inject(stellar::XDRInputFileStream&) override;
+    static void generateFuzzerInput(const std::string& filename);
```
`generateFuzzerInput` and `isBadInput`should be non-member non-friend functions for better encapsulation - just make them free functions inside of a namespace, e.g. FuzzUtils. (by the way, you can read up here on why this is a better practice: http://www.drdobbs.com/cpp/how-non-member-functions-improve-encapsu/184401197)

robertDurst replied:
Thanks for the link!

marta-lokhova commented on src/test/FuzzerImpl.h:
```diff
@@ -0,0 +1,43 @@
+  public:
+    OverlayFuzzer();
+    void inject(stellar::XDRInputFileStream&) override;
+    static void generateFuzzerInput(const std::string& filename);
```
Same comment for `generateFuzzerInput ` and `isBadInput`

robertDurst commented on src/test/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,366 @@
+
+    // crank until nodes are connected
+    mSimulation->crankUntil(
+        [simulation = mSimulation, acceptorIndex = ACCEPTOR_INDEX,
```
Ah yes, skimmed *Effective Modern C++* chapter on this the other day, but on closer look here there is no chance of dangling reference.

robertDurst commented on src/test/fuzz.cpp:
```diff
@@ -1,144 +1,63 @@
+createFuzzer(int processID, FuzzerMode fuzzerMode)
 {
-    try
+    switch (fuzzerMode)
```
So originally I defaulted to `TRANSACTION` mode, however, above in one of the collapsed comments, Nicolas advised against this: *"no silent failures please: if not recognized -> help*".

Open to either.

marta-lokhova replied:
what I mean here is that if a user doesn't provide a mode, default to overlay. If a user does provide a mode, but it's invalid, don't run the fuzzer, print help message. I don't see any benefit of having a fuzzer in unknown mode (we can't even do fuzzing with it, plus it's error-prone).

robertDurst commented:
Ok @jonjove and @marta-lokhova, I believe I fixed up the code to adhere to your feedback. Two places I am not 100% sure is:
* for db `clearPreparedCache`, I had this in two places. I removed the first one since I believe it makes most sense at the end, or at least as much sense as the beginning -- I also could not quite figure out why it had an affect on stability when placed before `XDRStreamInput` like it did; placed at the end it not only clears the statements between loops, but also makes stability happy.
* for the `--mode` cli flag, I have duplicated code for parsing the fuzzerMode input. However I did this because I kept running into lambda capture by reference dangling reference errors when I tried to extract these to a function (*this way of passing non-const references to multiple lambdas does not seem good, but its seems to be the way to do this based on the other command parsers*).

I also cleaned up my sporadic commits. The first commit that addressed both y'all's feedback is (_13be3d1f3fd2ff34f82fbe4ec252db3415bfa2d7_) or the first one dated July 24.

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,358 @@
+        Operation op = gen(FUZZER_INITIAL_CORPUS_OPERATION_GEN_SIZE);
+        while (isBadTransactionFuzzerInput(op))
+        {
+            op = gen(FUZZER_INITIAL_CORPUS_OPERATION_GEN_SIZE);
```
same comment than for overlay: don't use `FUZZER_INITIAL_CORPUS_OPERATION_GEN_SIZE` here but some higher number

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,358 @@
+    attemptApplication(Application& app, AbstractLedgerTxn& ltx)
+    {
+        // seed randomness
+        seedRandomness();
```
calls to `seedRandomness` and `resetForFuzzer` don't belong here, should be done in `inject`

MonsieurNicolas replied:
actually, should probably be moved in some method that can be called from multiple places `resetTxInternalState`

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,358 @@
+
+        // attempt application of transaction without accounting for sequence
+        // number, processing the fee, or committing the LedgerTxn
+        mCachedAccount.reset();
```
why do you need to reset `mCachedAccount` ? Don't we want to let it do its thing depending on protocol version?

robertDurst replied:
This was part of the [`TransactionFrame::apply` method](https://github.com/stellar/stellar-core/blob/master/src/transactions/TransactionFrame.cpp#L655) I recreated here. I believe this reset for `mCachedAccount` seems to happen regardless of protocol.

MonsieurNicolas replied:
I asked because `loadSourceAccount` seems to already do the right thing

robertDurst replied:
Yes, good point -- this was for a large portion of the fee processing and signature checking I left out. Removed.

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,358 @@
+    VirtualClock clock;
+    mApp = createTestApplication(clock, getFuzzConfig(processID));
+
+    LedgerTxn ltx(mApp->getLedgerTxnRoot());
```
you need to call `resetTxInternalState` before creating `ltx`

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,358 @@
+
+        try
+        {
+            // attempt to apply transaction
```
should call `resetTxInternalState` before creating ltx

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,358 @@
+        }
+        catch (...)
+        {
+            throw std::runtime_error(fmt::format("Unexpected internal error"));
```
(loss of context)

robertDurst replied:
No reason. Will fix. Is the following bad practice?
```
catch(...) {
   throw;
}
```

MonsieurNicolas replied:
if you're not doing anything, yes it's bad practice as it doesn't achieve anything

robertDurst replied:
Ok, makes sense -- therefore, this try catch in itself is not needed since we just propagate all exceptions.

robertDurst commented on src/test/fuzz.cpp:
```diff
@@ -147,84 +62,64 @@ fuzz(std::string const& filename, el::Level logLevel,
 
 void
-genfuzz(std::string const& filename)
+genfuzz(std::string const& filename, FuzzerMode fuzzerMode)
```
Not quite sure I follow here. 

Currently each fuzzer has a corresponding free function in the `FuzzerImpl` file with generation logic, that shares `isBadInput` with the fuzzer. These generation methods must either be free or static, because we do not need (and do not want) to initialize a fuzzer and start up apps just to create a random XDR. That being said, I could separate the initialization of the fuzzer to a separate method, and out of the constructor but I am not sure if this is the best route.

MonsieurNicolas replied:
yes, that's what I would like to see (switch on `mode` is basically what polymorphism does, in a cleaner way):
that way, each fuzzing mode is encapsulated in a dedicated class, if you move what is currently in the constructor into a method you'll have what you want.

robertDurst replied:
This should be addressed in my latest commits.

MonsieurNicolas commented on src/test/FuzzerImpl.cpp:
```diff
@@ -0,0 +1,360 @@
+        txFramePtr->attemptApplication(*mApp, ltx);
+
+        // ensure everything reset as ltx goes out of scope
+        resetTxInternalState(*mApp);
```
ltx is still in scope here, you need to do something like
```c++
{
   LedgerTxn ltx(mApp->getLedgerTxnRoot());
   txFramePtr->attemptApplication(*mApp, ltx);
}
resetTxInternalState(*mApp);
```

MonsieurNicolas commented:
r+ a9468013d6a4c955dba8580bba96789803698946

robertDurst commented:
@MonsieurNicolas accidentally turned off format for a commit, fixed now

MonsieurNicolas commented:
r+ 60b38b41535e1d31c092ecbe633a92bab03f9659

robertDurst commented:
My apologies -- the fuzzer utilized a method behind FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION. Fixed.

MonsieurNicolas commented:
r+ c2cd2fc752aa68dedf3043999f1872c666b89a5e
